### ISSUE TITLE
Port DS2490 code from libusb 0.1 to libusb 1.0

### ISSUE THREAD

**Issue by rfinnie** [bcl/digitemp#14]
From https://bugs.debian.org/810407

```
digitemp has a build-depends on libusb-dev. A few years ago upstream
has released a new major version libusb 1.0 with a different API which
aims to fix design deficiencies with USB 2.0 and 3.0 in mind.

The old libusb 0.1 package is not supported upstream anymore and should
be considered deprecated.

If digitemp supports the new libusb 1.0 library, please consider
switching the build-depends from libusb-dev to libusb-1.0-0-dev. If not
please inform upstream that porting the software to the new API is
recommended.
```

I've taken a brief look at it looks like porting the ds2490 stuff to libusb 1.0 would not be hugely difficult fo someone with libusb experience, but the work is probably beyond my C abilities.  Hopefully someone can pick this up.

**Comment by bcl** [comment]
http://libusb.sourceforge.net/api-1.0/index.html

**Comment by robert-scheck** [comment]
@bcl, may I kindly ask for your attention here? [Rename libusb packages and deprecate old API](https://fedoraproject.org/wiki/Changes/Rename_libusb_packages_and_deprecated_old_api) is on the list for Fedora 34…and I am not sure whether I would like to maintain a RPM package in Fedora with legacy requirements.

**Comment by bcl** [comment]
Yeah, I've looked into porting it and really don't feel like doing it myself. I actually use serial to usb adapters with iButtonLink devices here.

For Fedora I'd suggest dropping the build of the ds2490 version and just build the serial binaries.

**Comment by rct** [comment]
(Sorry Off Topic)  Are you using the iButtonLink masters in the mode where they emulate the Dallas masters?   

I think I saw the iButtonLink devices are supposed to have some advantages via their native protocol, but I've never seen any software that uses it.

**Comment by bcl** [comment]
Yes, using digitemp_DS9097U with them. Their ascii interface lets you have more control, but even without it  they work better with longer networks.

**Comment by rfinnie** [comment]
> Yeah, I've looked into porting it and really don't feel like doing it myself. I actually use serial to usb adapters with iButtonLink devices here.
> 
> For Fedora I'd suggest dropping the build of the ds2490 version and just build the serial binaries.

I'm now doing this for Debian as of version 3.7.2-2 in sid today (I wasn't getting immediate pressure to do so, but happened to check the [drop libusb 0.1 tracking bug](https://bugs.debian.org/810470) and saw there weren't many blockers left, so it was probably a matter of time).

Looks like rawhide has not yet dropped DS2490.

**Comment by haku** [comment]
i was able to get digitemp_DS2490 running on both debian 12 and nixos 26.05 with no code changes.
the trick is to compile it against the libusb 0.1 compatibility api.

## debian

```shell
$ apt install libusb-0.1-4 libusb-dev
$ make ds2490
$ sudo tee /etc/modprobe.d/blacklist-ds2490.conf <<EOF
blacklist ds2490
blacklist w1_therm
EOF
```
then reboot for modprobe config to take effect and it should work.

## nixos

shell.nix:
```nix
{ pkgs ? import <nixpkgs> {} }:
let
  digitemp = pkgs.digitemp.overrideAttrs (old: {
    makeFlags = [ "ds2490" ];
    buildInputs = [
      pkgs.libusb-compat-0_1
    ];
  });
in
pkgs.mkShell {
  packages = [ digitemp ];
}
```

will also need to stop the kernel stealing the device, so put this somewhere in system config, switch and reboot:
```nix
boot.blacklistedKernelModules = [
    "ds2490"
    "w1_smem"
    "w1_therm"
];
```

though on nixos it complains a bit:
```
$ sudo digitemp_DS2490 -q -a   
Jul 22 16:52:57 Sensor 0 C: 25.38 F: 77.68
libusb: warning [libusb_exit] device 4.1 still referenced
libusb: warning [libusb_exit] device 3.3 still referenced
libusb: warning [libusb_exit] device 3.2 still referenced
libusb: warning [libusb_exit] device 3.4 still referenced
libusb: warning [libusb_exit] device 3.1 still referenced
libusb: warning [libusb_exit] device 2.1 still referenced
libusb: warning [libusb_exit] device 1.1 still referenced
```

i guess some small fixes around releasing resources are needed?
but this seems a much smaller task than trying to port the whole thing.

**Comment by haku** [comment]
although....
all of that is maybe pointless cos at least on nixos 26.05 the kernel module also works:

```
fae@selkie:~  20260722-171119
$ lsusb                                          
Bus 003 Device 004: ID 04fa:2490 Dallas Semiconductor DS1490F 2-in-1 Fob, 1-Wire adapter

$ cat /sys/bus/w1/devices/28-000001f032af/temperature
25687
```
